libursa/src/bn/rust.rs

```diff
@@ -0,0 +1,632 @@
+        if n.bn.is_one() ||
+           n.bn.is_zero() {
+            return Err(UrsaCryptoError::InvalidStructure("Invalid modulus".to_string()))
+        }
+
+        // Euclid's extended algorithm, Bèzout coefficient of `n` is not needed
+        //n is either prime or coprime
+        //
+        //function inverse(a, n)
+        //    t := 0;     newt := 1;
+        //    r := n;     newr := a;
+        //    while newr ≠ 0
+        //        quotient := r div newr
+        //        (t, newt) := (newt, t - quotient * newt)
+        //        (r, newr) := (newr, r - quotient * newr)
+        //    if r > 1 then return "a is not invertible"
+        //    if t < 0 then t := t + n
+        //    return t
+        //
+        let (mut t, mut new_t) = (BigInt::zero(), BigInt::one());
+        let (mut r, mut new_r) = (n.bn.clone(), self.bn.clone());
+
+        while !new_r.is_zero() {
+            let quotient = &r / &new_r;
+            let temp_t = t.clone();
+            let temp_new_t = new_t.clone();
+
+            t = temp_new_t.clone();
+            new_t = temp_t - &quotient * temp_new_t;
+
+            let temp_r = r.clone();
+            let temp_new_r = new_r.clone();
+
+            r = temp_new_r.clone();
+            new_r = temp_r - quotient * temp_new_r;
+        }
+        if r > BigInt::one() {
+            return Err(UrsaCryptoError::InvalidStructure("Not invertible".to_string()));
+        } else if t < BigInt::zero() {
+            t += n.bn.clone()
+        }
+
+        Ok(BigNumber { bn: t })
+    }
+
+    pub fn set_negative(&self, negative: bool) -> Result<BigNumber, UrsaCryptoError> {
+        match (self.bn < BigInt::zero(), negative) {
+            (true, true) => Ok(BigNumber { bn: self.bn.clone() }),
+            (false, false) => Ok(BigNumber { bn: self.bn.clone() }),
+            (true, false) => Ok(BigNumber { bn: -self.bn.clone() }),
+            (false, true) => Ok(BigNumber { bn: -self.bn.clone() }),
+        }
+    }
+
+    pub fn is_negative(&self) -> bool {
+        self.bn.is_negative()
+    }
+
+    pub fn increment(&self) -> Result<BigNumber, UrsaCryptoError> {
+        Ok(BigNumber { bn: &self.bn + 1 })
+    }
+
+    pub fn decrement(&self) -> Result<BigNumber, UrsaCryptoError> {
+        Ok(BigNumber { bn: &self.bn - 1 })
+    }
+
+    pub fn lshift1(&self) -> Result<BigNumber, UrsaCryptoError> {
+        Ok(BigNumber { bn: &self.bn << 1 })
+    }
+
+    pub fn rshift1(&self) -> Result<BigNumber, UrsaCryptoError> {
+        Ok(BigNumber { bn: &self.bn >> 1 })
+    }
+
+    pub fn rshift(&self, n: i32) -> Result<BigNumber, UrsaCryptoError> {
+        let n = n as usize;
+        Ok(BigNumber { bn: &self.bn >> n })
+    }
+
+    pub fn mod_div(&self, b: &BigNumber, p: &BigNumber, _ctx: Option<&mut BigNumberContext>) -> Result<BigNumber, UrsaCryptoError> {
+        //(a * (1/b mod p) mod p)
+        let res = (&self.bn * b.inverse(p, _ctx)?.bn) % &p.bn;
```

Comment:
`p` can be negative, better to add the negative check in `modulus` and use the method.